Use client-go dynamic client instead of generated

stevesloka opened the pull request:
Updates #2113 

This switches from using the generated client for custom CRD types to the dynamic. Conversion from `Unstructured` to types objects happens in `internal/contour/handler.go` so all other logic afterward stays consistent. 

Signed-off-by: Steve Sloka <slokas@vmware.com>

stevesloka commented:
@youngnick @jpeach @davecheney I'd like to add some tests around `Unstructured` to validate that workflow, but I think this is ready for an initial look.

jpeach commented:
What happens to `deepcopy-gen` here? We use `DeepCopy` in `internal/k8s/status.go`, and this PR doesn't delete the `zz_generated.deepcopy.go` files, but it does delete the script that runs `deepcopy-gen`.

stevesloka commented on cmd/contour/serve.go:
```diff
@@ -179,11 +196,19 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
+	ingressRouteInformer := dynamicFactory.ForResource(schema.GroupVersionResource{Group: "contour.heptio.com", Version: "v1beta1", Resource: "ingressroutes"})
+	tlscertIRInformer := dynamicFactory.ForResource(schema.GroupVersionResource{Group: "contour.heptio.com", Version: "v1beta1", Resource: "tlscertificatedelegations"})
+	httpProxyInformer := dynamicFactory.ForResource(schema.GroupVersionResource{Group: "projectcontour.io", Version: "v1", Resource: "httpproxies"})
+	tlscertProxyInformer := dynamicFactory.ForResource(schema.GroupVersionResource{Group: "projectcontour.io", Version: "v1", Resource: "tlscertificatedelegations"})
```
Actually we have this already in the crd, "register.go", so can remove this.

stevesloka commented on cmd/contour/serve.go:
```diff
@@ -396,3 +425,13 @@ func startInformer(inf informer, log logrus.FieldLogger) func(stop <-chan struct
+
+func startDynamicInformer(inf cache.SharedIndexInformer, log logrus.FieldLogger) func(stop <-chan struct{}) error {
+	return func(stop <-chan struct{}) error {
+		log.Println("starting")
```
The informers are different types and have different ways to start. `informer` uses `Start()` and `cache.SharedIndexInformer` uses `Run()`. I'm not sure we can combine into the same method.

jpeach replied:
Ok, this part is extremely confusing.

The `informer` interface actually matches on the `coreinformers.SharedInformerFactory` type, because that type implements `(*internalinterfaces.SharedInformerFactory.Start)`. So we refer to both informers and informer factory as "informers" in the helper functions without being really clear about the differences.

So `dynamicFactory` is an object of type `*dynamic informer.DynamicSharedInformerFactory`, and it is this object that you want to `Start()` using `startInformer(). You only need to start the factory, not each informer that the factory generates.

We can improve this further (probably in a separate PR) by calling `WaitForCacheSync` on the factories rather than on each individual informer.

stevesloka commented on internal/contour/handler.go:
```diff
@@ -104,6 +122,80 @@ func (e *EventHandler) Start() func(<-chan struct{}) error {
 	return e.run
 }
 
+func (e *EventHandler) ConvertUnstructured(obj interface{}) interface{} {
```
Yeah good point, I bounced around this idea while building. I'll move this out.

stevesloka commented on internal/contour/handler.go:
```diff
@@ -19,17 +19,28 @@ package contour
 
+// Group Version for projectcontour.io/v1
+var projectcontourv1GV = schema.GroupVersion{Group: "projectcontour.io", Version: "v1"}
+var ingressroutev1beta1GV = schema.GroupVersion{Group: "contour.heptio.com", Version: "v1beta1"}
```
Thanks for pointing this out, I moved to use this, I think it makes sense to keep this constant here with the CRD spec.

stevesloka commented on internal/contour/handler.go:
```diff
@@ -104,6 +122,80 @@ func (e *EventHandler) Start() func(<-chan struct{}) error {
+					WithField("kind", obj.GetKind()).
+					WithField("version", obj.GetAPIVersion()).
+					Error(err)
+				return nil
```
After moving to k8s package, I change this to return an error so the caller can manage.

stevesloka commented on internal/k8s/status_test.go:
```diff
@@ -1,115 +0,0 @@
-// Copyright © 2019 VMware
```
They use a `fake` client from the client-go project, now that we're using a dynamic client, there isn't a matching fake client.

stevesloka commented:
> What happens to deepcopy-gen here? We use DeepCopy in internal/k8s/status.go, and this PR doesn't delete the zz_generated.deepcopy.go files, but it does delete the script that runs deepcopy-gen.

I added the same code the controller-runtime uses to generate the DeepCopy methods in the makefile.

stevesloka commented on internal/contour/handler_test.go:
```diff
@@ -0,0 +1,279 @@
+	run(t, "proxytlscertunstructured", testcase{
+		obj:  proxyTLSCertUnstructured,
+		want: proxyTLSCert1,
+	})
```
I added an unknown type. I'm struggling to find a way to test that it errors with a known type.

jpeach commented on internal/k8s/converter.go:
```diff
@@ -0,0 +1,82 @@
+		// Setup converter to understand custom CRD types
+		_ = projectcontour.AddKnownTypes(c.scheme)
+		_ = ingressroutev1.AddKnownTypes(c.scheme)
+	}
```
Nit: probably a bit cleaner to pull this into a `NewConverter()` API.

stevesloka replied:
> suggestion: if you don't care about the result, you doin't have to assign it to anything, ie

Yup I had that originally, but the linter complains about it if you don't handle the error. 

```
internal/k8s/converter.go:39:31: Error return value of `projectcontour.AddKnownTypes` is not checked (errcheck)
                projectcontour.AddKnownTypes(c.scheme)
                                            ^
internal/k8s/converter.go:40:31: Error return value of `ingressroutev1.AddKnownTypes` is not checked (errcheck)
                ingressroutev1.AddKnownTypes(c.scheme)
                                            ^
make: *** [Makefile:102: check-golint] Error 1

davecheney replied:
personally I think this is incorrect, the linter needs to shut up, but do what you think is appropriate.

stevesloka replied:
I agree with you, let me see if there's a way around this.

jpeach replied:
FWIW, you can silence the linter in a specific case with a [comment](https://github.com/golangci/golangci-lint#nolint) `// nolint(errcheck)`. Maybe in this case, we should check the error and panic.

davecheney replied:
yeah, you're right, if adding these types to the schema can reasonably fail then we should check the error (hint, check(err) is in scope in this package).

stevesloka replied:
I cleaned up the calling code to remove the `error`, it was always returning `nil`, so felt just as weird to check if error.

jpeach commented:
> On 7 Feb 2020, at 6:07 am, Steve Sloka <notifications@github.com> wrote:
> 
> I added an unknown type. I'm struggling to find a way to test that it errors with a known type.

What if you provide an Unstructured object of a know type where a known field is of the wrong type. For example, make "spec" an integer rather than a map/

jpeach commented on cmd/contour/serve.go:
```diff
@@ -179,11 +196,19 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
+	dynamicFactory := dynamicinformer.NewFilteredDynamicSharedInformerFactory(dynamicClient, 0, v1.NamespaceAll, nil)
+
+	dynamicClient.Resource(schema.GroupVersionResource{Group: "projectcontour.io", Version: "v1", Resource: "httpproxies"})
+	ingressRouteInformer := dynamicFactory.ForResource(schema.GroupVersionResource{Group: "contour.heptio.com", Version: "v1beta1", Resource: "ingressroutes"})
```
Nit: the `GroupVersionResource` is used in two places, so it might be worth hoisting into constants in each API package. Your call, if you think it's worth it.

davecheney commented on go.mod:
```diff
@@ -26,7 +26,7 @@ require (
 	gopkg.in/alecthomas/kingpin.v2 v2.2.6
 	gopkg.in/yaml.v2 v2.2.2
-	k8s.io/api v0.0.0-20190918195907-bd6ac527cfd2
+	k8s.io/api v0.0.0-20191003000013-35e20aa79eb8
```
Can we do this in two steps please.

1. Update the k8s client and api types, see also #2155 
2. Then do this PR.

stevesloka replied:
I submitted #2214 & #2213 which address these two points.

davecheney commented on internal/contour/handler.go:
```diff
@@ -65,6 +65,9 @@ type EventHandler struct {
 	seq int
+
+	// Converter handles conversion from unstructured to a type
+	converter k8s.Converter
```
I kinda feel that the field and the type both have missed the opportunity to obviate the comment. If k8s.Converter was called k8s.UnstructuredConverter then maybe the commend wouldn't be necessary.

davecheney commented on internal/contour/handler.go:
```diff
@@ -80,14 +83,34 @@ type opDelete struct {
 }
 
 func (e *EventHandler) OnAdd(obj interface{}) {
+	obj, err := e.converter.ConvertUnstructured(obj)
```
Is it possible to pull the e.Error change out into its own change, this looks like a worthwhile change in and of itself.

stevesloka replied:
Hey, not following what you are asking? The `e.Error` is new from the converter logic.

stevesloka commented on internal/k8s/converter.go:
```diff
@@ -0,0 +1,82 @@
+		case "HTTPProxy":
+			proxy := &projectcontour.HTTPProxy{}
+			err := c.scheme.Convert(obj, proxy, nil)
+			if err != nil {
```
good point! :+1:

stevesloka commented:
> What if you provide an Unstructured object of a know type where a known field is of the wrong type. For example, make "spec" an integer rather than a map/

@jpeach I added a case for this by swapping around the port/name in a service. Thanks for the suggestion! =)

stevesloka commented:
I think is ready for another look for you all, I've addressed all the review comments. 

@jpeach I moved the GroupResourceVersions up to where they were already declared in the apis. Thanks for that suggestion!

@davecheney I have a quick question on your comment here if you could clarify a bit for me:  https://github.com/projectcontour/contour/pull/2191#discussion_r376835126

davecheney commented on internal/k8s/converter.go:
```diff
@@ -0,0 +1,75 @@
+// Any other type passed other than unstructured is just returned back to caller
+func (c *UnstructuredConverter) Convert(obj interface{}) (interface{}, error) {
+	switch obj := obj.(type) {
+	case *unstructured.Unstructured:
```
I think this out switch can be rewritten as
```
unstructured, ok := obj.(*unstructured.Unstructured)
if !ok {
    return nil, fmt.Errorf("unsupported object type: %v", reflect.TypeOf(obj))
}
switch unstructured.GetKind() {
...
}
```
to reduce the overall indentation in this method

stevesloka replied:
This changes up the original idea of what this method did which I think is better. Before you could pass it a known type or an unstructured and it would return, but now it only converts an unstructured. This caused a check in the handler to only apply to the unstructured types.

davecheney commented on internal/contour/handler.go:
```diff
@@ -80,14 +82,34 @@ type opDelete struct {
 }
 
 func (e *EventHandler) OnAdd(obj interface{}) {
+	obj, err := e.UnstructuredConverter.Convert(obj)
```
its sort of a bit inconsistent that we refer to an embedded field by its fully qualified name, yet two lines below a different embedded object is called without qualification. I think it might be easier on the reader if either both were referred to by their embedded field names, or both anonymously.

davecheney commented on internal/contour/handler.go:
```diff
@@ -65,6 +65,8 @@ type EventHandler struct {
 	// an event has been received.
 	seq int
+
+	*k8s.UnstructuredConverter
```
Maybe we can decouple this a little with an anonymous interface eg
```
Converter interface{
     Convert(obj interface{}) (interface{}, error)
}
```

youngnick commented on internal/k8s/status.go:
```diff
@@ -168,7 +173,8 @@ func (irs *StatusWriter) setIngressRouteStatus(existing, updated *ingressroutev1
 	}
 
-	_, err = irs.Client.ContourV1beta1().IngressRoutes(existing.GetNamespace()).Patch(existing.GetName(), types.MergePatchType, patchBytes)
+	irGVR := schema.GroupVersionResource{Group: "contour.heptio.com", Version: "v1beta1", Resource: "ingressroutes"}
```
Shouldn't this use `ingressroutev1.IngressRouteGVR`?

youngnick commented on internal/k8s/status.go:
```diff
@@ -190,6 +196,7 @@ func (irs *StatusWriter) setHTTPProxyStatus(existing, updated *projcontour.HTTPP
 	}
 
-	_, err = irs.Client.ProjectcontourV1().HTTPProxies(existing.GetNamespace()).Patch(existing.GetName(), types.MergePatchType, patchBytes)
+	proxyGVR := schema.GroupVersionResource{Group: "projectcontour.io", Version: "v1", Resource: "httpproxies"}
```
I think this should also use `projcontour.HTTPProxyGVR` here.
